src/main/java/ch/jalu/configme/migration/VersionMigrationService.java

```diff
@@ -0,0 +1,236 @@
+     */
+    protected void handleTooLowerVersion(int version) {
+        throw new IllegalStateException("The current version (" + version + ") is lower than " + getStartVersion());
+    }
+
+    /**
+     * Invoked if the version is greater than {@link #getCurrentVersion()}.
+     */
+    protected void handleTooHigherVersion(int version) {
+        throw new IllegalStateException("The current version (" + version + ") is greater than " + getCurrentVersion());
+    }
+
+    /**
+     * Perform the migration by using the versioning system.
+     * <p>
+     * Note that the settings manager automatically saves the resource
+     * if the migration service returns {@link #MIGRATION_REQUIRED} from {@link #checkAndMigrate}.
+     *
+     * @param reader the reader with which the configuration file can be read
+     * @param configurationData the configuration data
+     * @return true if a migration has been performed, false otherwise (see constants on {@link MigrationService})
+     */
+    protected boolean performMigrations(@NotNull PropertyReader reader, @NotNull ConfigurationData configurationData) {
+        int currentConfigVersion = readConfigurationVersion(reader, configurationData);
+        AtomicBoolean migrationResult = new AtomicBoolean(NO_MIGRATION_NEEDED);
+
+        // Migrate the configuration from the version 1 to 2 to 3 to ...
+        migrations
+            .forEach(migration -> {
+                if (migration == null) return; // null-safe check.
+
+                int fromVersion = migration.fromVersion();
+                int toVersion = migration.toVersion();
+
+                // start the migration.
+                if (currentConfigVersion == fromVersion) {
+                    migration.migrate(reader, configurationData);
+                    setConfigurationVersion(configurationData, toVersion);
+                    migrationResult.set(MIGRATION_REQUIRED);
+                }
+            });
+
+        return migrationResult.get();
+    }
+
+    /**
+     * @return the nullable config version property comment.
+     */
+    @Nullable
+    protected String getConfigVersionPropertyComment() {
+        throw new IllegalStateException("Not implemented yet");
+        // return null;
+    }
+
+    /**
+     * Gets the current configuration version.
+     * If the property doesn't exist it will be created with the {@link #getStartVersion()}.
+     *
+     * @param reader the reader with which the configuration file can be read
+     * @param configurationData the configuration data
+     * @return the current configuration version.
+     */
+    protected int readConfigurationVersion(@NotNull PropertyReader reader, @NotNull ConfigurationData configurationData) {
+        Property<Integer> configVersionProperty = PropertyInitializer.newProperty(getVersionPropertyName(), getStartVersion());
+
+        // if the version property does not exist just add it.
+        if (!reader.contains(getVersionPropertyName())) {
```

Comment:
Interesting to note that this could (should?) trigger a rewrite of the config if the property doesn't exist, but it's a bit annoying to communicate that from this method since it already returns something. Probably fine like this, since any migration that is then triggered would trigger a rewrite.

However, I'm not sure if the missing version property should then be created with the _start version_ as value, because all bets are off if there's no version—not sure trying to run _all_ migrations is better than not doing anything... keeping in mind that default values will be used wherever stuff is missing. I would probably set the version config to the current application value.